### PR TITLE
fix: expand k8s pod discovery regex

### PR DESCRIPTION
**Pull request by kruskall**
update k8s pod ID discovery regex to the current status of the spec.

See https://github.com/elastic/apm/blob/main/specs/agents/metadata.md#containerkubernetes-metadata

Add a testcase to make sure the pod mentioned in the related issue is
discovered correctly.

Closes https://github.com/elastic/apm-agent-go/issues/822

### CODE REVIEW

File: internal/apmhostutil/container_linux.go [hunk 1]
```diff
@@ -46,12 +46,12 @@ var (
...
-			`(?:^/kubepods\.slice/kubepods-[^/]+\.slice/kubepods-[^/]+-pod([^/]+)\.slice/$)`,
+			`(?:kubepods[^/]*-pod([^/]+)\.slice/)`,
 	)
 
 	containerIDRegexp = regexp.MustCompile(
 		"^" +
-			"[[:xdigit:]]{64}|" +
+			"[[:xdigit:]]{64}$|" +
 			"[[:xdigit:]]{8}-[[:xdigit:]]{4}-[[:xdigit:]]{4}-[[:xdigit:]]{4}-[[:xdigit:]]{4,}" +
 			"$",
```
Comment (axw) [inline internal/apmhostutil/container_linux.go:56]: I think the original intention was that the "^" and "$" were to apply regardless of which of the other lines matched. Possibly there was originally a non-capturing group surrounding those options. Anyway, I think we should add "^" and "$" to each of the lines, so it's:

```go
	containerIDRegexp = regexp.MustCompile(
			"^[[:xdigit:]]{64}$|" +
			"^[[:xdigit:]]{8}-[[:xdigit:]]{4}-[[:xdigit:]]{4}-[[:xdigit:]]{4}-[[:xdigit:]]{4,}$",
	)
```

WDYT?

**Review by axw — COMMENTED** [review]
Should we also add the test case from the description of https://github.com/elastic/apm/pull/508?

File: internal/apmhostutil/container_linux.go (hunk 1 above)
Comment (kruskall) [reply internal/apmhostutil/container_linux.go]: I agree, this should also be more in line with the spec 👍

**Comment by kruskall** [comment]
It turns out that the test case in the description was not recognized as the discovery algorithm was outdated. I've updated the code, added comments and a test case 🏓

File: internal/apmhostutil/container_linux.go [hunk 2]
```diff
@@ -99,32 +97,40 @@ func readCgroupContainerInfo(r io.Reader) (*model.Container, *model.Kubernetes,
...
+				continue
+			}
+		}
+
+		dirname, basename := cgroupPath[:idx], cgroupPath[idx+1:]
+
+		// If the basename ends with ".scope", check for a hyphen and remove everything up to
+		// and including that. This allows us to match .../docker-<container-id>.scope as well
+		// as .../<container-id>.
+		if strings.HasSuffix(basename, ".scope") {
```
Comment (axw) [inline internal/apmhostutil/container_linux.go:126]: I think we can remove the `systemdScopeSuffix` constant?

**Review by axw — APPROVED** [review]

File: internal/apmhostutil/container_linux.go (hunk 2 above)
Comment (kruskall) [reply internal/apmhostutil/container_linux.go:122]: Good catch!

**Pull request merged**